### PR TITLE
O2-1203 rANS library for entropy coding

### PR DESCRIPTION
**Pull request by MichaelLettrich**
This adds a simple custom implementation of the rANS entropy coder to O2 which compresses data close to entropy level at high bandwidths.

### CODE REVIEW

**Comment by MichaelLettrich** [comment]
@matthiasrichter, @davidrohr: I added this under Utilities since the entropy coder is something shared. However i noticed that the huffman coder of @matthiasrichter  is already in there under "DataCompression". How should we proceed here?

**Comment by davidrohr** [comment]
I have never done anything in the Utilities/DataCompression folder, so @matthiasrichter should suggest what to do.

File: Utilities/rANS/test/test_ransEncodeDecode.cxx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  rans::SymbolStatistics stats{std::begin(SOURCE), std::end(SOURCE)};
+  stats.rescaleFrequencyTable(PROBABILITY_SCALE);
+
+  std::vector<stream_t> encoderBuffer(1 << 20, 0);
+  rans::Encoder<coder_t> encoder{stats, PROBABILITY_BITS};
+  auto ransBegin = encoder.process(encoderBuffer.begin(), encoderBuffer.end(), std::begin(SOURCE), std::end(SOURCE));
+
+  std::vector<source_t> decoderBuffer(stats.getMessageLength(), 0);
+  rans::Decoder<coder_t, source_t> decoder(stats, PROBABILITY_BITS, PROBABILITY_SCALE);
+  decoder.process(decoderBuffer.begin(), ransBegin, stats.getMessageLength());
```
Comment (shahor02) [inline Utilities/rANS/test/test_ransEncodeDecode.cxx:54]: perhaps I am missing something, but should not one decode starting from the encoded data, i.e. ``encoderBuffer``.
I think it would be useful to put some comments/guidelines on the steps in this test, e.g. what drives the particular choice of ``PROBABILITY_BITS``, ``stream_t`` etc. 
Is there an example of producing compressed data and storing it with its dictionary (e.g. dumping to a root file, as we discussed)?

**Comment by MichaelLettrich** [comment]
@shahor02 this PR was made to "bring over" the library from my own repo into o2 and adapt it to match style, formatting etc so that it will pass the CI and code chekers. IO functionality will be merged as a second part.

File: Utilities/rANS/test/test_ransEncodeDecode.cxx (hunk 1 above)
Comment (MichaelLettrich) [reply Utilities/rANS/test/test_ransEncodeDecode.cxx:54]: Thanks for the review. The code is correct but maybe not straight forward to understand. I will improve documentation -  Constants are explained in my CHEP Poster and paper. Plan to improve documentation as soon as the API is more stable.

Comment (MichaelLettrich) [reply Utilities/rANS/test/test_ransEncodeDecode.cxx]: I added some more documentation to the unittest to make sure it is easier to understand and linked my chep poster in the Readme.

**Comment by MichaelLettrich** [comment]
the library is now residing in the `o2::rans` namespace. hope that resolves all comments. Anything from your side @matthiasrichter ?

File: Utilities/rANS/test/test_ransEncodeDecode.cxx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  // By convention the encoder runs backwards over the input so that the decoder can return
+  // the data in the expected order. The encoded message ranges from ransBegin to encoderBuffer.end()
+  auto ransBegin = encoder.process(encoderBuffer.begin(), encoderBuffer.end(), std::begin(T::source), std::end(T::source));
+
+  // The decoded message will go into the decoder buffer which will have as many symbols as the original message
+  std::vector<typename T::source_t> decoderBuffer(std::end(T::source) - std::begin(T::source), 0);
+  // create a stateful decoder object that build a decoding table from the given symbol statistics
+  o2::rans::Decoder<typename T::coder_t, typename T::source_t> decoder(stats, T::probabilityBits, T::probabilityScale);
+  // the deocder unwinds the rANS state in the encoder buffer starting at ransBegin and decodes it into the decode buffer;
+  decoder.process(decoderBuffer.begin(), ransBegin, stats.getMessageLength());
```
Comment (shahor02) [inline Utilities/rANS/test/test_ransEncodeDecode.cxx:91]: OK, thanks! I would be more clear if you would start decoding explicitly from the encoded message rather than from intermediate iterator.

File: Utilities/rANS/include/librans/Coder.h [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    *r = ((x / freq) << scale_bits) + (x % freq) + start;
+  };
+
+  // Flushes the rANS encoder.
+  static void encFlush(State<T>* r, Stream_t** pptr)
+  {
+    T x = *r;
+    Stream_t* ptr = *pptr;
+
+    if constexpr (needs64Bit<T>()) {
```
Comment (matthiasrichter) [inline Utilities/rANS/include/librans/Coder.h:100]: I don't fully understand the logic here, this decision is purely on the state variable, but doesn't this require a consistency check for `Stream_t` as well? In the true branch it needs to be 4 byte wide, in the false branch 1 byte.

File: Utilities/rANS/include/librans/Coder.h [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+    *r = x;
+  };
+
+  // Encodes a given symbol. This is faster than straight RansEnc since we can do
+  // multiplications instead of a divide.
+  //
+  // See Rans32EncSymbolInit for a description of how this works.
+  static void encPutSymbol(State<T>* r, Stream_t** pptr, EncoderSymbol<T> const* sym, uint32_t scale_bits)
+  {
+#ifdef DEBUG
```
Comment (matthiasrichter) [inline Utilities/rANS/include/librans/Coder.h:167]: think this can be removed, `assert` is deactivated with the `NODEBUG` flag automatically.

Comment (MichaelLettrich) [reply Utilities/rANS/include/librans/Coder.h:167]: Thank you for the remark, will remove the ifdefs.

File: Utilities/rANS/include/librans/Coder.h (hunk 3 above)
Comment (MichaelLettrich) [reply Utilities/rANS/include/librans/Coder.h:100]: Thanks, indeed this makes no sense. There have originally been two different files one for 32 and one for 64 bit states, where they differed it seems that I blindly introduced the ` if constexpr (needs64Bit<T>())` to differenciate between the implementations. I will review them.

Comment (MichaelLettrich) [reply Utilities/rANS/include/librans/Coder.h:100]: So indeed the code is not universal but only works with the combination
64 bit arithmetic, 32 bit streaming, 32 bit arithmetic, 8 bit streaming. I really forgot about it because it has been a longer time since I last touched this.
To enforce correct types during compile time I now added a static assert into the `Coder` class. Obviously there are other possibilities and one can think of generalizing the `Coder` further, but it will require some more development and testing to ensure all corner cases are caught. 
For what we do 64Bit arithmeric and 32 bit streaming seems to be the most performant option anyhow.

File: Utilities/rANS/test/test_ransEncodeDecode.cxx (hunk 2 above)
Comment (MichaelLettrich) [reply Utilities/rANS/test/test_ransEncodeDecode.cxx:91]: Thanks for the comment. The problem is that I can't know how large the encoded message will come out of the encoder and therefore I have to allocate a buffer that is "big enough".  Since the rANS encoder works backwards it will iterate backwards over the buffer until all symbols were encoded the position where it stops is `ransBegin` - so the actual encoded message goes from `ransBegin`to `encoderBuffer.end()`. It takes another copy to an adequatly sized buffer to pass this in a nicer way. I will document this better and improve the naming to be more obvious.

**Comment by MichaelLettrich** [comment]
Thanks for all the good feedback. I will merge this as soon as it passes the tests again.

**Comment by MichaelLettrich** [comment]
@davidrohr @ktf  what is the matter with the GPU builds - seems they are installing again but failing with some of the tests?

**Comment by shahor02** [comment]
Shall we merge this? What is next, when can we expect an API?

**Comment by davidrohr** [comment]
+1 for merging this
